<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Two Factor Genome Wide Association Study</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<link rel="stylesheet" href="/opt/R/x86_64-pc-linux-gnu-library/4.0/rmarkdown/rmarkdown/templates/html_vignette/resources/vignette.css" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">Two Factor Genome Wide Association Study</h1>



<p>The Two Factor model is an extension of the One Factor Model. The difference between the one and two factor models are the additional regression coefficient, and the correlations between the two latent variables.</p>
<p>Unlike the One Factor GWAS model, the Two Factor GWAS model does not have an analogue in the current GWAS literature but nevertheless plays an integral role the many genetic theories of comorbidity and pleiotropy. For example, multivariate twin models routinely detect genetic variation across a range of phenotypes, and interpret these findings in terms of genetic variants that are shared between the phenotypes. LD score regression analyses take this one step further, testing for genetic overlap between phenotypes based on summary statistics from univariate GWAS. The two factor GWAS model takes this theoretical hypothesis further than existing methods by allowing users to explicitly test for pleiotropy in a single model. This framework allows users to think about phenotypes from an explicitly multivariate perspective that cannot be achieved with alternative GWAS software. A general schematic depiction of the two factor model is presented in the figure below, but users should keep in mind that the number of items in the analysis and the specific parameterization of the model can be adjusted to suit the specific research question.</p>
<p>In the Figure below, the latent factor (F<sub>1</sub>) is the underlying (unmeasured) cause the observed items (x<sub>k</sub>). The association between the latent factor and the observed indicators, (x<sub>1</sub>, x<sub>2</sub> &amp; x<sub>3</sub>), are estimated by the factor loadings (λ<sub>1</sub>, λ<sub>2</sub>, &amp; λ<sub>3</sub>, respectively). The residual variances (δ<sub>1</sub>, δ<sub>2</sub>, &amp; δ<sub>3</sub>) indicate the remaining variance in (x<sub>k</sub>) that is not shared with the latent factor. The regression of the latent factor on the SNP (for all SNPs in the analysis) is depicted by (β<sub>F1</sub>). The corresponding factor loadings for the second latent factor (F<sub>1</sub>) and the y<sub>k</sub> variables can be interpreted similarly. The new parameter that is added to the two factor model that was absent from the one factor model is the covariance between the latent variables (ψ). This parameter can be interpreted in conjunction with the β<sub>F1</sub> and β<sub>F2</sub> parameters to assess the impact of the genomic associations on comorbidity.</p>
<p>For the two factor model to be identified, you will need at least three items for each factor (as was the case with the one factor model). It is a good idea to use more items than the minimum; more items will improve the measurement accuracy of the latent variables.</p>
<div class="figure">
<img src="https://jpritikin.github.io/gwsem/exampleFigs/twoFactor.jpg" alt="A schematic depiction of the Two-Factor GWAS Model" />
<p class="caption">A schematic depiction of the Two-Factor GWAS Model</p>
</div>
<div id="example" class="section level2">
<h2>Example</h2>
<p>To demonstrate how to conduct a Two Factor GWAS, we use the same simulated GWAS data for a one factor GWAS model with 6,000 individuals consisting of 2,000 SNPs, and simulated novel phenotypic data with six items (A<sub>1</sub>, A<sub>2</sub> &amp; A<sub>3</sub> for F<sub>1</sub>, and B<sub>1</sub>, B<sub>2</sub> &amp; B<sub>3</sub> for F<sub>2</sub>) as well as 6 covariates (that are a proxy for age, sex, ancestry principle components or other confounds). The latent variables F<sub>1</sub> and F<sub>2</sub> were simultaneously regressed on each SNP.</p>
</div>
<div id="running-the-two-factor-gwas" class="section level2">
<h2>Running the Two-Factor GWAS</h2>
<p>The first step is opening R and loading the GW-SEM package into the R computing environment (which will also load all of the dependencies, such as OpenMx), which can be achieved by using the command below:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="kw">library</span>(gwsem)</a>
<a class="sourceLine" id="cb1-2" title="2"><span class="co">#&gt; Loading required package: OpenMx</span></a>
<a class="sourceLine" id="cb1-3" title="3"><span class="co">#&gt; To take full advantage of multiple cores, use:</span></a>
<a class="sourceLine" id="cb1-4" title="4"><span class="co">#&gt;   mxOption(key=&#39;Number of Threads&#39;, value=parallel::detectCores()) #now</span></a>
<a class="sourceLine" id="cb1-5" title="5"><span class="co">#&gt;   Sys.setenv(OMP_NUM_THREADS=parallel::detectCores()) #before library(OpenMx)</span></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-7" title="7"><span class="co">#&gt; Attaching package: &#39;gwsem&#39;</span></a>
<a class="sourceLine" id="cb1-8" title="8"><span class="co">#&gt; The following object is masked from &#39;package:base&#39;:</span></a>
<a class="sourceLine" id="cb1-9" title="9"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb1-10" title="10"><span class="co">#&gt;     signif</span></a></code></pre></div>
<p>The next step is to load the phenotypic data. This will include any items that define the latent variables or necessary covariates that you will include in the analysis, such as age, sex, or ancestry principle components. Once the phenotypic data has been loaded, it is advisable to make sure the data has been loaded properly by, e.g., inspecting the first few lines of data, looking at the means and variances, etc. This can be done with the code below:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1">location &lt;-<span class="st"> &#39;https://jpritikin.github.io/gwsem/gwsemTwoFacExample&#39;</span></a>
<a class="sourceLine" id="cb2-2" title="2">TwoDat &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">file.path</span>(location, <span class="st">&quot;phenoTwoData.txt&quot;</span>), <span class="dt">header=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb2-3" title="3"><span class="kw">head</span>(TwoDat)</a>
<a class="sourceLine" id="cb2-4" title="4"><span class="co">#&gt;           A1          A2         A3          B1         B2        B3</span></a>
<a class="sourceLine" id="cb2-5" title="5"><span class="co">#&gt; 1 0.99530058  0.73647170  1.8979517  0.16796993 0.48912621 -1.151701</span></a>
<a class="sourceLine" id="cb2-6" title="6"><span class="co">#&gt; 2 1.69508013  2.97669551  2.5059896  0.60355319 0.09443592  0.885876</span></a>
<a class="sourceLine" id="cb2-7" title="7"><span class="co">#&gt; 3 0.48069152  0.07470696  0.2401195  0.26878160 1.37390510 -0.348631</span></a>
<a class="sourceLine" id="cb2-8" title="8"><span class="co">#&gt; 4 0.43838397  1.46105679  0.2129391  2.85912185 1.46017993  2.545553</span></a>
<a class="sourceLine" id="cb2-9" title="9"><span class="co">#&gt; 5 1.58112970 -0.65526690 -0.2898117 -0.08739975 1.08364671  1.392842</span></a>
<a class="sourceLine" id="cb2-10" title="10"><span class="co">#&gt; 6 0.05250986  1.23666372 -0.9969216  0.76547539 1.16942874  0.359331</span></a>
<a class="sourceLine" id="cb2-11" title="11"><span class="co">#&gt;           pc1          pc2        pc3         pc4         pc5</span></a>
<a class="sourceLine" id="cb2-12" title="12"><span class="co">#&gt; 1 -1.73534591 -0.005949438 -0.5308992  0.76801845 -1.50932522</span></a>
<a class="sourceLine" id="cb2-13" title="13"><span class="co">#&gt; 2  0.47007030 -0.127011839  0.7976675 -0.64860921  0.05576587</span></a>
<a class="sourceLine" id="cb2-14" title="14"><span class="co">#&gt; 3  1.33173044  1.480261629 -1.6033764 -0.04209832 -0.57337829</span></a>
<a class="sourceLine" id="cb2-15" title="15"><span class="co">#&gt; 4 -0.07900914  0.097746365 -1.2246490  0.97834643  0.14235702</span></a>
<a class="sourceLine" id="cb2-16" title="16"><span class="co">#&gt; 5  0.63493316  1.940078145  1.5124544 -0.32060878  0.16354672</span></a>
<a class="sourceLine" id="cb2-17" title="17"><span class="co">#&gt; 6 -0.62710577  0.051160950  0.9269276 -0.02472826  2.79142175</span></a></code></pre></div>
<p>Once the data is loaded into R, you can recode the data, transform it, and tell R if we have ordinal or binary indicators using <code>mxFactor()</code>. The data for this example are simulated and continuous, and therefore, we will not be doing anything now, but if you would like to chop the indicators up into binary or ordinal variables, this would be when you would do it.</p>
<p>After you are satisfied the data is in the appropriate shape we can build a two factor GWAS model with the following command:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1"></a>
<a class="sourceLine" id="cb3-2" title="2">twoFac &lt;-<span class="st"> </span><span class="kw">buildTwoFac</span>(<span class="dt">phenoData =</span> TwoDat,                                         <span class="co"># what the data object is (which you read in above)</span></a>
<a class="sourceLine" id="cb3-3" title="3">                     <span class="dt">F1itemNames =</span> <span class="kw">c</span>(<span class="st">&quot;A1&quot;</span>, <span class="st">&quot;A2&quot;</span>, <span class="st">&quot;A3&quot;</span>),                <span class="co"># what the items of the latent factor are</span></a>
<a class="sourceLine" id="cb3-4" title="4">                     <span class="dt">F2itemNames =</span> <span class="kw">c</span>(<span class="st">&quot;B1&quot;</span>, <span class="st">&quot;B2&quot;</span>, <span class="st">&quot;B3&quot;</span>),                <span class="co"># what the items of the latent factor are</span></a>
<a class="sourceLine" id="cb3-5" title="5">                     <span class="dt">covariates=</span><span class="kw">c</span>(<span class="st">&#39;pc1&#39;</span>,<span class="st">&#39;pc2&#39;</span>,<span class="st">&#39;pc3&#39;</span>,<span class="st">&#39;pc4&#39;</span>,<span class="st">&#39;pc5&#39;</span>),                    <span class="co"># what covariates that you want to include in the analysis</span></a>
<a class="sourceLine" id="cb3-6" title="6">                     <span class="dt">fitfun =</span> <span class="st">&quot;WLS&quot;</span>,</a>
<a class="sourceLine" id="cb3-7" title="7">                     <span class="dt">exogenous =</span> T)                                                 <span class="co"># and the fit function that you would like to use (WLS is much faster than ML)</span></a></code></pre></div>
<p>You can take this object (addFac) which is technically an OpenMx model, and fit it using <code>mxRun()</code></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">twoFacFit &lt;-<span class="st"> </span><span class="kw">mxRun</span>(twoFac)</a>
<a class="sourceLine" id="cb4-2" title="2"><span class="co">#&gt; Running TwoFac with 53 parameters</span></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="kw">summary</span>(twoFacFit)</a>
<a class="sourceLine" id="cb4-4" title="4"><span class="co">#&gt; Summary of TwoFac </span></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="co">#&gt;  </span></a>
<a class="sourceLine" id="cb4-6" title="6"><span class="co">#&gt; free parameters:</span></a>
<a class="sourceLine" id="cb4-7" title="7"><span class="co">#&gt;            name matrix row col      Estimate   Std.Error lbound ubound</span></a>
<a class="sourceLine" id="cb4-8" title="8"><span class="co">#&gt; 1     snp_to_F1      A  F1 snp  0.0204846184 0.020988048              </span></a>
<a class="sourceLine" id="cb4-9" title="9"><span class="co">#&gt; 2     snp_to_F2      A  F2 snp  0.0576206101 0.021378949              </span></a>
<a class="sourceLine" id="cb4-10" title="10"><span class="co">#&gt; 3  F1_lambda_A1      A  A1  F1  0.7346342607 0.013551599              </span></a>
<a class="sourceLine" id="cb4-11" title="11"><span class="co">#&gt; 4  F1_lambda_A2      A  A2  F1  0.7379333831 0.013631470              </span></a>
<a class="sourceLine" id="cb4-12" title="12"><span class="co">#&gt; 5  F1_lambda_A3      A  A3  F1  0.7377062366 0.013901195              </span></a>
<a class="sourceLine" id="cb4-13" title="13"><span class="co">#&gt; 6  F2_lambda_B1      A  B1  F2  0.7303859008 0.013596925              </span></a>
<a class="sourceLine" id="cb4-14" title="14"><span class="co">#&gt; 7  F2_lambda_B2      A  B2  F2  0.7271160639 0.013485047              </span></a>
<a class="sourceLine" id="cb4-15" title="15"><span class="co">#&gt; 8  F2_lambda_B3      A  B3  F2  0.7322454226 0.013603676              </span></a>
<a class="sourceLine" id="cb4-16" title="16"><span class="co">#&gt; 9     pc1_to_A1      A  A1 pc1 -0.0098844413 0.013253671              </span></a>
<a class="sourceLine" id="cb4-17" title="17"><span class="co">#&gt; 10    pc1_to_A2      A  A2 pc1 -0.0028782404 0.013078169              </span></a>
<a class="sourceLine" id="cb4-18" title="18"><span class="co">#&gt; 11    pc1_to_A3      A  A3 pc1 -0.0024240569 0.012972208              </span></a>
<a class="sourceLine" id="cb4-19" title="19"><span class="co">#&gt; 12    pc1_to_B1      A  B1 pc1  0.0186090846 0.013413044              </span></a>
<a class="sourceLine" id="cb4-20" title="20"><span class="co">#&gt; 13    pc1_to_B2      A  B2 pc1  0.0055690079 0.013444384              </span></a>
<a class="sourceLine" id="cb4-21" title="21"><span class="co">#&gt; 14    pc1_to_B3      A  B3 pc1  0.0286550850 0.013081062              </span></a>
<a class="sourceLine" id="cb4-22" title="22"><span class="co">#&gt; 15    pc2_to_A1      A  A1 pc2 -0.0233794780 0.013537836              </span></a>
<a class="sourceLine" id="cb4-23" title="23"><span class="co">#&gt; 16    pc2_to_A2      A  A2 pc2  0.0057159064 0.013110043              </span></a>
<a class="sourceLine" id="cb4-24" title="24"><span class="co">#&gt; 17    pc2_to_A3      A  A3 pc2 -0.0223660926 0.013434802              </span></a>
<a class="sourceLine" id="cb4-25" title="25"><span class="co">#&gt; 18    pc2_to_B1      A  B1 pc2 -0.0052081586 0.013003029              </span></a>
<a class="sourceLine" id="cb4-26" title="26"><span class="co">#&gt; 19    pc2_to_B2      A  B2 pc2  0.0094819900 0.013148629              </span></a>
<a class="sourceLine" id="cb4-27" title="27"><span class="co">#&gt; 20    pc2_to_B3      A  B3 pc2  0.0135662703 0.012938699              </span></a>
<a class="sourceLine" id="cb4-28" title="28"><span class="co">#&gt; 21    pc3_to_A1      A  A1 pc3  0.0127788665 0.013261193              </span></a>
<a class="sourceLine" id="cb4-29" title="29"><span class="co">#&gt; 22    pc3_to_A2      A  A2 pc3  0.0047702053 0.013254165              </span></a>
<a class="sourceLine" id="cb4-30" title="30"><span class="co">#&gt; 23    pc3_to_A3      A  A3 pc3 -0.0026084923 0.013391766              </span></a>
<a class="sourceLine" id="cb4-31" title="31"><span class="co">#&gt; 24    pc3_to_B1      A  B1 pc3  0.0054261628 0.013078376              </span></a>
<a class="sourceLine" id="cb4-32" title="32"><span class="co">#&gt; 25    pc3_to_B2      A  B2 pc3 -0.0027204869 0.013280429              </span></a>
<a class="sourceLine" id="cb4-33" title="33"><span class="co">#&gt; 26    pc3_to_B3      A  B3 pc3 -0.0082655820 0.013399066              </span></a>
<a class="sourceLine" id="cb4-34" title="34"><span class="co">#&gt; 27    pc4_to_A1      A  A1 pc4  0.0134712533 0.012907348              </span></a>
<a class="sourceLine" id="cb4-35" title="35"><span class="co">#&gt; 28    pc4_to_A2      A  A2 pc4 -0.0073100816 0.013191213              </span></a>
<a class="sourceLine" id="cb4-36" title="36"><span class="co">#&gt; 29    pc4_to_A3      A  A3 pc4  0.0098127266 0.013130184              </span></a>
<a class="sourceLine" id="cb4-37" title="37"><span class="co">#&gt; 30    pc4_to_B1      A  B1 pc4 -0.0085092911 0.013096330              </span></a>
<a class="sourceLine" id="cb4-38" title="38"><span class="co">#&gt; 31    pc4_to_B2      A  B2 pc4 -0.0244776304 0.013156055              </span></a>
<a class="sourceLine" id="cb4-39" title="39"><span class="co">#&gt; 32    pc4_to_B3      A  B3 pc4 -0.0303754950 0.013213978              </span></a>
<a class="sourceLine" id="cb4-40" title="40"><span class="co">#&gt; 33    pc5_to_A1      A  A1 pc5  0.0076189162 0.013123408              </span></a>
<a class="sourceLine" id="cb4-41" title="41"><span class="co">#&gt; 34    pc5_to_A2      A  A2 pc5  0.0211575205 0.013375857              </span></a>
<a class="sourceLine" id="cb4-42" title="42"><span class="co">#&gt; 35    pc5_to_A3      A  A3 pc5  0.0026325210 0.013428760              </span></a>
<a class="sourceLine" id="cb4-43" title="43"><span class="co">#&gt; 36    pc5_to_B1      A  B1 pc5  0.0004494821 0.012912829              </span></a>
<a class="sourceLine" id="cb4-44" title="44"><span class="co">#&gt; 37    pc5_to_B2      A  B2 pc5  0.0110870638 0.012814509              </span></a>
<a class="sourceLine" id="cb4-45" title="45"><span class="co">#&gt; 38    pc5_to_B3      A  B3 pc5  0.0015379813 0.012891719              </span></a>
<a class="sourceLine" id="cb4-46" title="46"><span class="co">#&gt; 39      snp_res      S snp snp  0.4914271980 0.012497156  0.001       </span></a>
<a class="sourceLine" id="cb4-47" title="47"><span class="co">#&gt; 40       A1_res      S  A1  A1  0.5088743473 0.014383598  0.001       </span></a>
<a class="sourceLine" id="cb4-48" title="48"><span class="co">#&gt; 41       A2_res      S  A2  A2  0.5083565302 0.014459753  0.001       </span></a>
<a class="sourceLine" id="cb4-49" title="49"><span class="co">#&gt; 42       A3_res      S  A3  A3  0.5094838535 0.014464749  0.001       </span></a>
<a class="sourceLine" id="cb4-50" title="50"><span class="co">#&gt; 43       B1_res      S  B1  B1  0.5094995862 0.014217909  0.001       </span></a>
<a class="sourceLine" id="cb4-51" title="51"><span class="co">#&gt; 44       B2_res      S  B2  B2  0.5106029750 0.014287910  0.001       </span></a>
<a class="sourceLine" id="cb4-52" title="52"><span class="co">#&gt; 45       B3_res      S  B3  B3  0.5083648346 0.014650971  0.001       </span></a>
<a class="sourceLine" id="cb4-53" title="53"><span class="co">#&gt; 46       facCov      S  F1  F2  0.3275215106 0.015546164              </span></a>
<a class="sourceLine" id="cb4-54" title="54"><span class="co">#&gt; 47      snpMean      M   1 snp  0.9892453763 0.009049099              </span></a>
<a class="sourceLine" id="cb4-55" title="55"><span class="co">#&gt; 48       A1Mean      M   1  A1  0.5289081056 0.020160666              </span></a>
<a class="sourceLine" id="cb4-56" title="56"><span class="co">#&gt; 49       A2Mean      M   1  A2  0.5277992917 0.020173286              </span></a>
<a class="sourceLine" id="cb4-57" title="57"><span class="co">#&gt; 50       A3Mean      M   1  A3  0.5283450317 0.020211257              </span></a>
<a class="sourceLine" id="cb4-58" title="58"><span class="co">#&gt; 51       B1Mean      M   1  B1  0.5001707056 0.020351738              </span></a>
<a class="sourceLine" id="cb4-59" title="59"><span class="co">#&gt; 52       B2Mean      M   1  B2  0.5008052998 0.020248122              </span></a>
<a class="sourceLine" id="cb4-60" title="60"><span class="co">#&gt; 53       B3Mean      M   1  B3  0.4989284534 0.020406069              </span></a>
<a class="sourceLine" id="cb4-61" title="61"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb4-62" title="62"><span class="co">#&gt; Model Statistics: </span></a>
<a class="sourceLine" id="cb4-63" title="63"><span class="co">#&gt;                |  Parameters  |  Degrees of Freedom  |  Fit (r&#39;Wr units)</span></a>
<a class="sourceLine" id="cb4-64" title="64"><span class="co">#&gt;        Model:             53                  41947             3.000038</span></a>
<a class="sourceLine" id="cb4-65" title="65"><span class="co">#&gt;    Saturated:             35                  41965             0.000000</span></a>
<a class="sourceLine" id="cb4-66" title="66"><span class="co">#&gt; Independence:             14                  41986                   NA</span></a>
<a class="sourceLine" id="cb4-67" title="67"><span class="co">#&gt; Number of observations/statistics: 6000/42000</span></a>
<a class="sourceLine" id="cb4-68" title="68"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb4-69" title="69"><span class="co">#&gt; chi-square:  χ² ( df=17 ) = 3.000038,  p = 0.9999305</span></a>
<a class="sourceLine" id="cb4-70" title="70"><span class="co">#&gt; CFI: NA </span></a>
<a class="sourceLine" id="cb4-71" title="71"><span class="co">#&gt; TLI: NA   (also known as NNFI) </span></a>
<a class="sourceLine" id="cb4-72" title="72"><span class="co">#&gt; RMSEA: 0 *(Non-centrality parameter is negative)  [95% CI (0, 0)]</span></a>
<a class="sourceLine" id="cb4-73" title="73"><span class="co">#&gt; Prob(RMSEA &lt;= 0.05): 1</span></a>
<a class="sourceLine" id="cb4-74" title="74"><span class="co">#&gt; To get additional fit indices, see help(mxRefModels)</span></a>
<a class="sourceLine" id="cb4-75" title="75"><span class="co">#&gt; timestamp: 2020-07-10 20:14:37 </span></a>
<a class="sourceLine" id="cb4-76" title="76"><span class="co">#&gt; Wall clock time: 0.5719059 secs </span></a>
<a class="sourceLine" id="cb4-77" title="77"><span class="co">#&gt; optimizer:  SLSQP </span></a>
<a class="sourceLine" id="cb4-78" title="78"><span class="co">#&gt; OpenMx version number: 2.17.4 </span></a>
<a class="sourceLine" id="cb4-79" title="79"><span class="co">#&gt; Need help?  See help(mxSummary)</span></a></code></pre></div>
<p>This is strongly advised, as it is a great time to test whether the model is being specified the way you want it to be and that you are not getting unexpected estimates.</p>
<p>Provided that the model looks reasonable, you can plug the model that you have built into the <code>GWAS</code> function using the command below:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="kw">library</span>(curl)</a>
<a class="sourceLine" id="cb5-2" title="2"><span class="kw">curl_download</span>(<span class="kw">file.path</span>(location, <span class="st">&#39;example.pgen&#39;</span>),</a>
<a class="sourceLine" id="cb5-3" title="3">              <span class="kw">file.path</span>(<span class="kw">tempdir</span>(),<span class="st">&#39;example.pgen&#39;</span>))</a>
<a class="sourceLine" id="cb5-4" title="4"><span class="kw">curl_download</span>(<span class="kw">file.path</span>(location, <span class="st">&#39;example.pvar&#39;</span>),</a>
<a class="sourceLine" id="cb5-5" title="5">              <span class="kw">file.path</span>(<span class="kw">tempdir</span>(),<span class="st">&#39;example.pvar&#39;</span>))</a>
<a class="sourceLine" id="cb5-6" title="6"></a>
<a class="sourceLine" id="cb5-7" title="7"><span class="kw">GWAS</span>(<span class="dt">model =</span> twoFac,                                                                 <span class="co"># what model object you would like to fit</span></a>
<a class="sourceLine" id="cb5-8" title="8">    <span class="dt">snpData =</span> <span class="kw">file.path</span>(<span class="kw">tempdir</span>(),<span class="st">&#39;example.pgen&#39;</span>),                                                        <span class="co"># that path to the snpData file.</span></a>
<a class="sourceLine" id="cb5-9" title="9">    <span class="dt">out=</span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;twoFac.log&quot;</span>))                                                                <span class="co"># the file that you would like to save the full results into</span></a>
<a class="sourceLine" id="cb5-10" title="10"><span class="co">#&gt; Running TwoFac with 53 parameters</span></a>
<a class="sourceLine" id="cb5-11" title="11"><span class="co">#&gt; Done. See &#39;/tmp/RtmprJh9JP/twoFac.log&#39; for results</span></a>
<a class="sourceLine" id="cb5-12" title="12">                                                                           <span class="co"># the index of the snps (how many) you would like to fit</span></a></code></pre></div>
<p><strong>Note about the snpData argument:</strong> The path to your snpData will likely include switching to a different directory (as you will likely do your analysis in a different folder than your SNP data). All you need to do is point to the data using relative paths. Further, it will accept plink bed/bim/fam or pgen/psam/pvar data or bgen data (Oxford format)</p>
<p><strong>Note about the SNP argument:</strong> This can be used to run a limited number of SNP (i.e. not the whole snp file). This is particularly useful if you would like to run chop up a chr into several parts without cutting you actual genotype data into separate files.</p>
<p>While the <code>GWAS</code> function will take a while and frequently be done on a computing cluster, it is very useful to run a few SNPs (say 10 or 50) in an interactive session to ensure that: all of your relative file paths to the genotypes are correct, the model is taking a reasonable amount of time (i.e. 1-2 seconds/snp), the SNPs are giving sensible estimates, etc., as the results from a few SNPs can often tell you if there is a problem and that you are running a nonsensical model.</p>
<p>Refer to local documentation to learn how to submit jobs to your cluster queue submission systems. The R script that you provide will typically be similar to the one below:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" title="1"><span class="kw">library</span>(gwsem)</a>
<a class="sourceLine" id="cb6-2" title="2"></a>
<a class="sourceLine" id="cb6-3" title="3">TwoDat &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="kw">file.path</span>(location, <span class="st">&quot;phenoTwoData.txt&quot;</span>), <span class="dt">header=</span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb6-4" title="4"><span class="kw">head</span>(TwoDat)</a>
<a class="sourceLine" id="cb6-5" title="5"><span class="co">#&gt;           A1          A2         A3          B1         B2        B3</span></a>
<a class="sourceLine" id="cb6-6" title="6"><span class="co">#&gt; 1 0.99530058  0.73647170  1.8979517  0.16796993 0.48912621 -1.151701</span></a>
<a class="sourceLine" id="cb6-7" title="7"><span class="co">#&gt; 2 1.69508013  2.97669551  2.5059896  0.60355319 0.09443592  0.885876</span></a>
<a class="sourceLine" id="cb6-8" title="8"><span class="co">#&gt; 3 0.48069152  0.07470696  0.2401195  0.26878160 1.37390510 -0.348631</span></a>
<a class="sourceLine" id="cb6-9" title="9"><span class="co">#&gt; 4 0.43838397  1.46105679  0.2129391  2.85912185 1.46017993  2.545553</span></a>
<a class="sourceLine" id="cb6-10" title="10"><span class="co">#&gt; 5 1.58112970 -0.65526690 -0.2898117 -0.08739975 1.08364671  1.392842</span></a>
<a class="sourceLine" id="cb6-11" title="11"><span class="co">#&gt; 6 0.05250986  1.23666372 -0.9969216  0.76547539 1.16942874  0.359331</span></a>
<a class="sourceLine" id="cb6-12" title="12"><span class="co">#&gt;           pc1          pc2        pc3         pc4         pc5</span></a>
<a class="sourceLine" id="cb6-13" title="13"><span class="co">#&gt; 1 -1.73534591 -0.005949438 -0.5308992  0.76801845 -1.50932522</span></a>
<a class="sourceLine" id="cb6-14" title="14"><span class="co">#&gt; 2  0.47007030 -0.127011839  0.7976675 -0.64860921  0.05576587</span></a>
<a class="sourceLine" id="cb6-15" title="15"><span class="co">#&gt; 3  1.33173044  1.480261629 -1.6033764 -0.04209832 -0.57337829</span></a>
<a class="sourceLine" id="cb6-16" title="16"><span class="co">#&gt; 4 -0.07900914  0.097746365 -1.2246490  0.97834643  0.14235702</span></a>
<a class="sourceLine" id="cb6-17" title="17"><span class="co">#&gt; 5  0.63493316  1.940078145  1.5124544 -0.32060878  0.16354672</span></a>
<a class="sourceLine" id="cb6-18" title="18"><span class="co">#&gt; 6 -0.62710577  0.051160950  0.9269276 -0.02472826  2.79142175</span></a>
<a class="sourceLine" id="cb6-19" title="19"></a>
<a class="sourceLine" id="cb6-20" title="20">twoFac &lt;-<span class="st"> </span><span class="kw">buildTwoFac</span>(<span class="dt">phenoData =</span> TwoDat,                              <span class="co"># data object</span></a>
<a class="sourceLine" id="cb6-21" title="21">                     <span class="dt">F1itemNames =</span> <span class="kw">c</span>(<span class="st">&quot;A1&quot;</span>, <span class="st">&quot;A2&quot;</span>, <span class="st">&quot;A3&quot;</span>),                <span class="co"># items of the first latent factor</span></a>
<a class="sourceLine" id="cb6-22" title="22">                     <span class="dt">F2itemNames =</span> <span class="kw">c</span>(<span class="st">&quot;B1&quot;</span>, <span class="st">&quot;B2&quot;</span>, <span class="st">&quot;B3&quot;</span>),                <span class="co"># items of the second latent factor </span></a>
<a class="sourceLine" id="cb6-23" title="23">                     <span class="dt">covariates=</span><span class="kw">c</span>(<span class="st">&#39;pc1&#39;</span>,<span class="st">&#39;pc2&#39;</span>,<span class="st">&#39;pc3&#39;</span>,<span class="st">&#39;pc4&#39;</span>,<span class="st">&#39;pc5&#39;</span>),      <span class="co"># covariates </span></a>
<a class="sourceLine" id="cb6-24" title="24">                     <span class="dt">fitfun =</span> <span class="st">&quot;WLS&quot;</span>,</a>
<a class="sourceLine" id="cb6-25" title="25">             <span class="dt">exogenous =</span> T)                                                 </a>
<a class="sourceLine" id="cb6-26" title="26"></a>
<a class="sourceLine" id="cb6-27" title="27"></a>
<a class="sourceLine" id="cb6-28" title="28"><span class="kw">GWAS</span>(<span class="dt">model =</span> twoFac,</a>
<a class="sourceLine" id="cb6-29" title="29">     <span class="dt">snpData =</span> <span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&#39;example.pgen&#39;</span>),</a>
<a class="sourceLine" id="cb6-30" title="30">     <span class="dt">out=</span><span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;twoFac.log&quot;</span>))</a>
<a class="sourceLine" id="cb6-31" title="31"><span class="co">#&gt; Running TwoFac with 53 parameters</span></a>
<a class="sourceLine" id="cb6-32" title="32"><span class="co">#&gt; Done. See &#39;/tmp/RtmprJh9JP/twoFac.log&#39; for results</span></a></code></pre></div>
<p>The next step is to read the results into R. The output from the <code>GWAS</code> function contains all of the estimates and standard errors for all parameters in the model (for each SNP), as well as other general model fitting information. While you are unlikely to want to read all of this output into R, it is useful to do this on a small test GWAS analysis to ensure that you are getting reasonable results and to familiarize yourself with what information is available, in case you need to explore a specific parameter or model in more detail. This can be done using the command below:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">TwoResult &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;twoFac.log&quot;</span>))</a></code></pre></div>
<p>More likely, you are going to want to read the results into R for specific coefficients that inform a particular hypothesis. For the two factor model, we might want to look at both regression paths from the SNP to each latent factor as well as the covariance between the latent factors for jointly significant SNPs. To read the data for all three parameters of interest into R we can use the <code>loadResults</code> function. This function can be repeated for as many parameters that you want to read into R. It takes takes two arguments: the path to the data and the column name from the results file for the parameter that you want to examine.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">succinct &lt;-<span class="st"> </span><span class="kw">loadResults</span>(<span class="dt">path =</span> <span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;twoFac.log&quot;</span>), <span class="dt">focus =</span> <span class="kw">paste0</span>(<span class="st">&quot;snp_to_F&quot;</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>))</a>
<a class="sourceLine" id="cb8-2" title="2">succinct1 &lt;-<span class="st"> </span><span class="kw">signif</span>(succinct, <span class="st">&quot;snp_to_F1&quot;</span>)</a>
<a class="sourceLine" id="cb8-3" title="3">succinct2 &lt;-<span class="st"> </span><span class="kw">signif</span>(succinct, <span class="st">&quot;snp_to_F2&quot;</span>)</a></code></pre></div>
<p>We can then construct a Manhattan plot using the GW-SEM plot() function to visually inspect the general pattern of associations between the SNPs and each factor.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw">plot</span>(succinct1)</a></code></pre></div>
<p><img src="TwoFactorModel_files/figure-html/unnamed-chunk-10-1.png" /><!-- --></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1"><span class="kw">plot</span>(succinct2)</a></code></pre></div>
<p><img src="TwoFactorModel_files/figure-html/unnamed-chunk-10-2.png" /><!-- --></p>
<p>For the Two Factor Model, we will likely want to examine pleiotropic genetic effects by looking at how the pattern of SNP regression affect the correlation between the latent factors. We would expect that if a single SNP was associated with both factors, we would see a reduction in the correlation between the factors. Accordingly, we can sort the SNP results that we read into R by the largest Z statistic to see whether there were any SNPs jointly predicted both latent factors.</p>
<p>To demonstrate the intricacies of the Two Factor GWAS model, when simulating the data, we randomly chose snp141 to be associated with both factors, snp50 to be associated with F<sub>1</sub> only, and snp719 to be associated with F<sub>2</sub> only.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="kw">head</span>(succinct1[<span class="kw">order</span>(succinct1<span class="op">$</span>Z, <span class="dt">decreasing =</span> T),])</a>
<a class="sourceLine" id="cb11-2" title="2"><span class="co">#&gt;    MxComputeLoop1 CHR   BP     SNP A1 A2 statusCode catch1  snp_to_F1</span></a>
<a class="sourceLine" id="cb11-3" title="3"><span class="co">#&gt; 1:            142   1  141  snp141  B  A         OK     NA 0.35513064</span></a>
<a class="sourceLine" id="cb11-4" title="4"><span class="co">#&gt; 2:             51   1   50   snp50  A  B         OK     NA 0.30405528</span></a>
<a class="sourceLine" id="cb11-5" title="5"><span class="co">#&gt; 3:            469   1  468  snp468  B  A         OK     NA 0.08488139</span></a>
<a class="sourceLine" id="cb11-6" title="6"><span class="co">#&gt; 4:           1245   1 1244 snp1244  B  A         OK     NA 0.07377304</span></a>
<a class="sourceLine" id="cb11-7" title="7"><span class="co">#&gt; 5:            133   1  132  snp132  A  B         OK     NA 0.07040013</span></a>
<a class="sourceLine" id="cb11-8" title="8"><span class="co">#&gt; 6:           1718   1 1717 snp1717  A  B         OK     NA 0.06241153</span></a>
<a class="sourceLine" id="cb11-9" title="9"><span class="co">#&gt;        snp_to_F2 Vsnp_to_F1:snp_to_F1 Vsnp_to_F2:snp_to_F2         Z</span></a>
<a class="sourceLine" id="cb11-10" title="10"><span class="co">#&gt; 1:  0.3220622714         0.0006427296         0.0006405601 14.007931</span></a>
<a class="sourceLine" id="cb11-11" title="11"><span class="co">#&gt; 2: -0.0005456362         0.0006603582         0.0006263756 11.832125</span></a>
<a class="sourceLine" id="cb11-12" title="12"><span class="co">#&gt; 3:  0.0251987798         0.0006209106         0.0006214755  3.406418</span></a>
<a class="sourceLine" id="cb11-13" title="13"><span class="co">#&gt; 4:  0.0139381304         0.0006123987         0.0006169231  2.981127</span></a>
<a class="sourceLine" id="cb11-14" title="14"><span class="co">#&gt; 5:  0.0197922928         0.0006348669         0.0006404336  2.794037</span></a>
<a class="sourceLine" id="cb11-15" title="15"><span class="co">#&gt; 6:  0.0081323251         0.0006193158         0.0006167869  2.507891</span></a>
<a class="sourceLine" id="cb11-16" title="16"><span class="co">#&gt;               P</span></a>
<a class="sourceLine" id="cb11-17" title="17"><span class="co">#&gt; 1: 1.394079e-44</span></a>
<a class="sourceLine" id="cb11-18" title="18"><span class="co">#&gt; 2: 2.663118e-32</span></a>
<a class="sourceLine" id="cb11-19" title="19"><span class="co">#&gt; 3: 6.582131e-04</span></a>
<a class="sourceLine" id="cb11-20" title="20"><span class="co">#&gt; 4: 2.871894e-03</span></a>
<a class="sourceLine" id="cb11-21" title="21"><span class="co">#&gt; 5: 5.205456e-03</span></a>
<a class="sourceLine" id="cb11-22" title="22"><span class="co">#&gt; 6: 1.214540e-02</span></a>
<a class="sourceLine" id="cb11-23" title="23"></a>
<a class="sourceLine" id="cb11-24" title="24"><span class="kw">head</span>(succinct2[<span class="kw">order</span>(succinct2<span class="op">$</span>Z, <span class="dt">decreasing =</span> T),])</a>
<a class="sourceLine" id="cb11-25" title="25"><span class="co">#&gt;    MxComputeLoop1 CHR   BP     SNP A1 A2 statusCode catch1    snp_to_F1</span></a>
<a class="sourceLine" id="cb11-26" title="26"><span class="co">#&gt; 1:            142   1  141  snp141  B  A         OK     NA  0.355130641</span></a>
<a class="sourceLine" id="cb11-27" title="27"><span class="co">#&gt; 2:            720   1  719  snp719  B  A         OK     NA  0.031123462</span></a>
<a class="sourceLine" id="cb11-28" title="28"><span class="co">#&gt; 3:           1836   1 1835 snp1835  B  A         OK     NA  0.046207990</span></a>
<a class="sourceLine" id="cb11-29" title="29"><span class="co">#&gt; 4:           1504   1 1503 snp1503  B  A         OK     NA  0.024659160</span></a>
<a class="sourceLine" id="cb11-30" title="30"><span class="co">#&gt; 5:           1943   1 1942 snp1942  A  B         OK     NA  0.027306283</span></a>
<a class="sourceLine" id="cb11-31" title="31"><span class="co">#&gt; 6:           1047   1 1046 snp1046  B  A         OK     NA -0.004855771</span></a>
<a class="sourceLine" id="cb11-32" title="32"><span class="co">#&gt;     snp_to_F2 Vsnp_to_F1:snp_to_F1 Vsnp_to_F2:snp_to_F2         Z            P</span></a>
<a class="sourceLine" id="cb11-33" title="33"><span class="co">#&gt; 1: 0.32206227         0.0006427296         0.0006405601 12.725062 4.291361e-37</span></a>
<a class="sourceLine" id="cb11-34" title="34"><span class="co">#&gt; 2: 0.29285060         0.0006145705         0.0006546552 11.445632 2.472942e-30</span></a>
<a class="sourceLine" id="cb11-35" title="35"><span class="co">#&gt; 3: 0.06551444         0.0006257175         0.0006267728  2.616869 8.874046e-03</span></a>
<a class="sourceLine" id="cb11-36" title="36"><span class="co">#&gt; 4: 0.06542593         0.0006245056         0.0006363121  2.593671 9.495741e-03</span></a>
<a class="sourceLine" id="cb11-37" title="37"><span class="co">#&gt; 5: 0.06326570         0.0006197235         0.0006243532  2.531938 1.134340e-02</span></a>
<a class="sourceLine" id="cb11-38" title="38"><span class="co">#&gt; 6: 0.06100967         0.0006305413         0.0006396459  2.412286 1.585283e-02</span></a></code></pre></div>
<p>As can be seen, snp141 significantly predicts both latent factors. Let’s take a look at this correlation between the latent factors for this SNP.</p>
<p>First, let’s read in the factor correlation results, specifically asking to also load the standard errors, and treating the data object as a data frame. This makes it easier to use base R functions to search through the results. Let us look at the factor correlation for snp141, as well as snp50 and snp719, and a few other snps (say the first 6 in the results file).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">facCov &lt;-<span class="st"> </span><span class="kw">loadResults</span>(<span class="dt">path =</span> <span class="kw">file.path</span>(<span class="kw">tempdir</span>(), <span class="st">&quot;twoFac.log&quot;</span>), <span class="dt">focus =</span>  <span class="st">&quot;facCov&quot;</span>)</a>
<a class="sourceLine" id="cb12-2" title="2"></a>
<a class="sourceLine" id="cb12-3" title="3"><span class="kw">subset</span>(facCov, SNP <span class="op">==</span><span class="st"> &quot;snp141&quot;</span>)</a>
<a class="sourceLine" id="cb12-4" title="4"><span class="co">#&gt;    MxComputeLoop1 CHR  BP    SNP A1 A2 statusCode catch1    facCov</span></a>
<a class="sourceLine" id="cb12-5" title="5"><span class="co">#&gt; 1:            142   1 141 snp141  B  A         OK     NA 0.3010986</span></a>
<a class="sourceLine" id="cb12-6" title="6"><span class="co">#&gt;    VfacCov:facCov</span></a>
<a class="sourceLine" id="cb12-7" title="7"><span class="co">#&gt; 1:   0.0002510583</span></a>
<a class="sourceLine" id="cb12-8" title="8"></a>
<a class="sourceLine" id="cb12-9" title="9"><span class="kw">subset</span>(facCov, SNP <span class="op">==</span><span class="st"> &quot;snp50&quot;</span>)</a>
<a class="sourceLine" id="cb12-10" title="10"><span class="co">#&gt;    MxComputeLoop1 CHR BP   SNP A1 A2 statusCode catch1    facCov VfacCov:facCov</span></a>
<a class="sourceLine" id="cb12-11" title="11"><span class="co">#&gt; 1:             51   1 50 snp50  A  B         OK     NA 0.3331035   0.0002418059</span></a>
<a class="sourceLine" id="cb12-12" title="12"></a>
<a class="sourceLine" id="cb12-13" title="13"><span class="kw">subset</span>(facCov, SNP <span class="op">==</span><span class="st"> &quot;snp719&quot;</span>)</a>
<a class="sourceLine" id="cb12-14" title="14"><span class="co">#&gt;    MxComputeLoop1 CHR  BP    SNP A1 A2 statusCode catch1    facCov</span></a>
<a class="sourceLine" id="cb12-15" title="15"><span class="co">#&gt; 1:            720   1 719 snp719  B  A         OK     NA 0.3293708</span></a>
<a class="sourceLine" id="cb12-16" title="16"><span class="co">#&gt;    VfacCov:facCov</span></a>
<a class="sourceLine" id="cb12-17" title="17"><span class="co">#&gt; 1:   0.0002384187</span></a>
<a class="sourceLine" id="cb12-18" title="18"></a>
<a class="sourceLine" id="cb12-19" title="19"><span class="kw">head</span>(facCov)</a>
<a class="sourceLine" id="cb12-20" title="20"><span class="co">#&gt;    MxComputeLoop1 CHR BP  SNP A1 A2 statusCode catch1    facCov VfacCov:facCov</span></a>
<a class="sourceLine" id="cb12-21" title="21"><span class="co">#&gt; 1:              1   1  0 snp0  B  A         OK     NA 0.3287116   0.0002406796</span></a>
<a class="sourceLine" id="cb12-22" title="22"><span class="co">#&gt; 2:              2   1  1 snp1  A  B         OK     NA 0.3289716   0.0002414917</span></a>
<a class="sourceLine" id="cb12-23" title="23"><span class="co">#&gt; 3:              3   1  2 snp2  B  A         OK     NA 0.3280331   0.0002410056</span></a>
<a class="sourceLine" id="cb12-24" title="24"><span class="co">#&gt; 4:              4   1  3 snp3  A  B         OK     NA 0.3279913   0.0002408202</span></a>
<a class="sourceLine" id="cb12-25" title="25"><span class="co">#&gt; 5:              5   1  4 snp4  A  B         OK     NA 0.3282823   0.0002410134</span></a>
<a class="sourceLine" id="cb12-26" title="26"><span class="co">#&gt; 6:              6   1  5 snp5  B  A         OK     NA 0.3280614   0.0002412557</span></a></code></pre></div>
<p>As can be seen, there is a marked reduction on the correlation between the two latent factors for snp141 that is not present for the other SNPs that are associated with only one of the factors, nor for SNPs that are not associated with either factor. Accordingly, we should interpret both the associations with the latent factors and the reduced correlation between the phenotypes when discussing the results.</p>
</div>
<div id="suggestions-for-conducting-the-analysis" class="section level2">
<h2>Suggestions for Conducting the Analysis</h2>
<p>Before running any sort of analysis in GW-SEM, users should have a general understanding of the latent traits that they are interested in examining. Specifically, for the one-factor model, the items that load onto the latent variable should be established prior to running the GWAS, or the user risks conducting an analysis that does not test their intended hypothesis. For example, if the structure of the latent factor differs from the model written in GW-SEM (model misspecification), any conclusions that are drawn from the analysis are questionable. Alternatively, if there are items that do not load onto the latent factor, this may dilute the GWAS signal. The best, and most interpretable, results will come from a clearly defined, well specified, phenotypic model.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
